src/app/qgsgeometryvalidationdock.cpp

```diff
@@ -0,0 +1,343 @@
+  mZoomToFeatureButton->setEnabled( hasFeature );
+
+  switch ( mLastZoomToAction )
+  {
+    case  ZoomToProblem:
+      zoomToProblem();
+      break;
+
+    case ZoomToFeature:
+      zoomToFeature();
+      break;
+  }
+}
+
+void QgsGeometryValidationDock::onCurrentLayerChanged( QgsMapLayer *layer )
+{
+  if ( layer == mCurrentLayer )
+    return;
+
+  if ( mCurrentLayer )
+  {
+    disconnect( mCurrentLayer, &QgsVectorLayer::editingStarted, this, &QgsGeometryValidationDock::onLayerEditingStatusChanged );
+    disconnect( mCurrentLayer, &QgsVectorLayer::editingStopped, this, &QgsGeometryValidationDock::onLayerEditingStatusChanged );
+  }
+
+  mCurrentLayer = qobject_cast<QgsVectorLayer *>( layer );
+
+  if ( mCurrentLayer )
+  {
+    connect( mCurrentLayer, &QgsVectorLayer::editingStarted, this, &QgsGeometryValidationDock::onLayerEditingStatusChanged );
+    connect( mCurrentLayer, &QgsVectorLayer::editingStopped, this, &QgsGeometryValidationDock::onLayerEditingStatusChanged );
+  }
+
+  onLayerEditingStatusChanged();
+}
+
+void QgsGeometryValidationDock::onLayerEditingStatusChanged()
+{
+  bool enabled = false;
+  if ( mCurrentLayer && mCurrentLayer->isSpatial() && mCurrentLayer->isEditable() )
+  {
+    const QList<QgsGeometryCheckFactory *> topologyCheckFactories = QgsAnalysis::instance()->geometryCheckRegistry()->geometryCheckFactories( mCurrentLayer, QgsGeometryCheck::LayerCheck, QgsGeometryCheck::Flag::AvailableInValidation );
+    const QStringList activeChecks = mCurrentLayer->geometryOptions()->geometryChecks();
+    for ( const QgsGeometryCheckFactory *factory : topologyCheckFactories )
+    {
+      if ( activeChecks.contains( factory->id() ) )
+      {
+        enabled = true;
+        break;
+      }
+    }
+  }
+  mTopologyChecksPendingButton->setEnabled( enabled );
+}
+
+void QgsGeometryValidationDock::showHighlight( const QModelIndex &current )
+{
+  QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( mMapCanvas->currentLayer() );
+  if ( vlayer )
+  {
+    QgsGeometry featureGeometry = current.data( QgsGeometryValidationModel::FeatureGeometryRole ).value<QgsGeometry>();
+    QgsGeometry errorGeometry = current.data( QgsGeometryValidationModel::ErrorGeometryRole ).value<QgsGeometry>();
+    QgsPointXY locationGeometry = current.data( QgsGeometryValidationModel::ErrorLocationGeometryRole ).value<QgsPointXY>();
+
+    mFeatureRubberband->setToGeometry( featureGeometry );
+
+    QPropertyAnimation *featureAnimation = new QPropertyAnimation( mFeatureRubberband, "fillColor" );
+    featureAnimation->setEasingCurve( QEasingCurve::OutQuad );
+    connect( featureAnimation, &QPropertyAnimation::finished, featureAnimation, &QPropertyAnimation::deleteLater );
+    connect( featureAnimation, &QPropertyAnimation::valueChanged, this, [this]
+    {
+      mFeatureRubberband->update();
+    } );
+
+    featureAnimation->setDuration( 2000 );
```

Comment:
That's pretty slow -- it's not too annoying?

Reply:
Maybe, but that's adjusted quickly and better reviewed on real tests than in the code.